clang/lib/Sema/SemaExpr.cpp

```diff
@@ -9188,6 +9187,27 @@ Sema::CheckAssignmentConstraints(QualType LHSType, ExprResult &RHS,
   LHSType = Context.getCanonicalType(LHSType).getUnqualifiedType();
   RHSType = Context.getCanonicalType(RHSType).getUnqualifiedType();
 
+  // __builtin_counted_by_ref cannot be assigned to a variable, used in
+  // function call, or in a return.
+  auto FindBuiltinCountedByRefExpr = [](Expr *E) {
+    struct BuiltinCountedByRefVisitor
+        : public RecursiveASTVisitor<BuiltinCountedByRefVisitor> {
+      MemberExpr *ME = nullptr;
+      bool VisitMemberExpr(MemberExpr *E) {
```

Comment:
I couldn't find a check to find the use of __builtin_counted_by_ref. Shouldn't it be instead `VisitCallExpr` to check if there is a call to the builtin on RHS?

Reply:
At this point, the call to `__builtin_counted_by_ref` is gone and we have only the assignment left. That's why I added the `isBoundsSafetyCounter` flag to the `MemberDecl`. There's probably a more elegant way to do this, but I wasn't able to come up with one.

Reply:
Oh I see. I somehow missed that you are replacing the call to builtin with AddrOf MemberDecl in `SemaChecking.cpp`. But then doesn't it prevent innocent cases like `size_t count = ptr->count;` and `ptr->count = abc;`?

Also, Clang owners may raise a concern that merely replacing the builtin with `Addrof(MemberExpr(count))` in Sema we are loosing the source information as written. IIRC, Clang always wants to keep the whole source info as written. @AaronBallman Or please let me know if this is acceptable.

I guess you had to do it because of the difficulty of correctly emitting the member access, right? In case @AaronBallman says that the current implementation is not acceptable in Clang, off top of my head, you may want to create a new expression in AST, like, BuiltinCountRefExpr which contains the member expr that you created as a subexpression like this:

`BuiltinCountRefExpr(ArgExpr, MemberExpr(count))`, like Clang creates `OffsetOfExpr` for builtin_offsetof.

Then, you will handle the expr in CodeGen to evaluate the member expr.

Reply:
> But then doesn't it prevent innocent cases like `size_t count = ptr->count;` and `ptr-> count = abc;`?

It won't, because the flag is only set through this builtin.

> BuiltinCountRefExpr(ArgExpr, MemberExpr(count)), like Clang creates OffsetOfExpr for builtin_offsetof

I'm fine with doing this, if that's the better option. As for location information, I fudge a bit and assign the location of `__builtin_counted_by_ref` to the `AddrOf(MemberExpr(count))`. So highly gross. But adding a new builtin class seems very heavy weight. There are several builtins that are evaluated during parsing and replaced with AST nodes without having to resort to adding a whole new `Expr`.

Reply:
Perhaps the builtin should be converted after the sema checking? I'm just not sure where...

Reply:
> It won't, because the flag is only set through this builtin.

Is it? It looks like ` CountFD->setBoundsSafetyCounter(true);` is called inside `Sema::CheckCountedByAttrOnField` so I thought the flag is always set for count fields. Am I missing something?

> BuiltinCountRefExpr(ArgExpr, MemberExpr(count)), like Clang creates OffsetOfExpr for builtin_offsetof
I'm fine with doing this, if that's the better option. As for location information, I fudge a bit and assign the location of __builtin_counted_by_ref to the AddrOf(MemberExpr(count)). So highly gross. But adding a new builtin class seems very heavy weight. There are several builtins that are evaluated during parsing and replaced with AST nodes without having to resort to adding a whole new Expr.

I think the main issue might be that it would break Clang AST's principle of "faithfulness" to the original source, which is necessary for refactoring tools to be able to reconstruct the source from the AST (see https://clang.llvm.org/docs/InternalsManual.html#faithfulness). That said, if other builtins are already doing that then it might still be fine. We should probably wait on @AaronBallman @rjmccall's sign off though.

Reply:
> I think the main issue might be that it would break Clang AST's principle of "faithfulness" to the original source, which is necessary for refactoring tools to be able to reconstruct the original source from the AST (see https://clang.llvm.org/docs/InternalsManual.html#faithfulness). That said, if other builtins are already doing that then it might still be fine. We should probably wait on @AaronBallman @rjmccall's sign off though.

I understand. If we go the route of adding a new `*Expr` class, we might want to explore adding a generic `BuiltinExpr` class for this and potentially future builtins that we want to convert during Sema.

Reply:
Friendly ping to @AaronBallman and @rjmccall for their opinions on whether to create a new AST node for this builtin or keep this code, which replaces the builtin with a direct reference to the "counter."

Reply:
I think we need to keep the builtin around because otherwise things like `-ast-print` and `-ast-dump` will print/dump the semantic form rather than the source form. Generally speaking, we'd leave the builtin as a call expression and then handle that call expression with special builtin semantics from places like the constant expression interpreter and codegen, so we may not actually need a new AST node to represent this.

Reply:
@AaronBallman I think the reason why @bwendling moved it to AST was that emitting IRs for the member access (without creating the MemberExpr in AST) was quite tricky because IIRC, CodeGen didn't have such helper functions to emit nested member expressions without `MemberExpr`. That was why I was suggesting to have some AST node to pass around the `MemberExpr`. @bwendling do you remember what was exactly the problem in doing it purely in IRGen (without creating AST nodes)?

Reply:
I suppose. Adding a new AST node is a massive undertaking. I'll see how far I can get with the `CallExpr`.